daemon/api_sideload_n_try.go

```diff
@@ -685,22 +685,35 @@ func readComponentInfo(st *state.State, upload *uploadedContainer, flags sideloa
 		return nil, nil, BadRequest("cannot retrieve information for %q: %v", instanceName, err)
 	}
 
-	csi, err := snapasserts.DeriveComponentSideInfo(compRef.ComponentName, upload.tmpPath, info, model, assertstate.DB(st))
+	db := assertstate.DB(st)
+
+	csi, err := snapasserts.DeriveComponentSideInfo(compRef.ComponentName, upload.tmpPath, info, model, db)
 	if err != nil {
-		if errors.Is(err, &asserts.NotFoundError{}) {
-			// with devmode we try to find assertions but it's ok if they
-			// are not there (implies --dangerous)
-			if !flags.DevMode {
-				msg := "cannot find signatures with metadata for snap/component"
-				if upload.filename != "" {
-					msg = fmt.Sprintf("%s %q", msg, upload.filename)
-				}
-				return nil, nil, BadRequest(msg)
+		if !errors.Is(err, &asserts.NotFoundError{}) {
+			return nil, nil, BadRequest(err.Error())
+		}
+
+		msg := "cannot find signatures with metadata for snap/component"
+		if upload.filename != "" {
+			msg = fmt.Sprintf("%s %q", msg, upload.filename)
+		}
+
+		if !flags.DevMode {
+			if upload.filename != "" {
+				msg = fmt.Sprintf("%s %q", msg, upload.filename)
 			}
+			return nil, nil, BadRequest(msg)
+		}
 
-			csi = snap.NewComponentSideInfo(compRef, snap.Revision{})
-		} else {
-			return nil, nil, BadRequest(err.Error())
+		csi = snap.NewComponentSideInfo(compRef, snap.Revision{})
+	}
+
+	// make sure that we've got a resource pair for this component and snap
+	// revision. installing via snapstate checks this too, but we might as well
+	// fail early.
+	if !flags.DevMode {
+		if err := checkForResourcePair(csi, info, db); err != nil {
```

Comment:
a slightly different approach would be do add a assertstate.SnapResourcPair function to assertstate?

Reply:
Thanks, done.